### PR TITLE
Preserve float64 log-softmax tail gradients

### LINKED ISSUE #126630: tf.nn.log_softmax loses a finite first-derivative signal for float64 input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.log_softmax` returns zero for one first-derivative component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

`log_softmax` is equivariant to a common logit shift, so the resulting input VJP should sum to zero. In the reproducer below, the expected gradient is approximately `[5.551115123125776e-17, -5.551115123125776e-17]`.

However, TensorFlow returns `[0.0, -5.551115123125776e-17]`. The first gradient component is incorrectly collapsed to zero, and the resulting VJP no longer preserves the expected zero-sum relation.

Expected behavior?

The first-order autodiff result for `tf.nn.log_softmax` should preserve the finite, normal, nonzero `float64` gradient component. For the given input and cotangent, the gradient should be close to `[5.551115123125776e-17, -5.551115123125776e-17]`, and the VJP should sum to zero.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

dtype = tf.float64

expected = [5.551115123125776e-17, -5.551115123125776e-17]

z = tf.Variable([37.42994775023705, 0.0], dtype=dtype)
cotangent = tf.constant([1.0, 0.0], dtype=dtype)

with tf.GradientTape() as tape:
    y = tf.nn.log_softmax(z, axis=-1)
    objective = tf.reduce_sum(y * cotangent)

actual = tape.gradient(objective, z)
vjp_sum = tf.reduce_sum(actual)

print("actual:")
print(actual.numpy().tolist())
print("expected:")
print(expected)
print("VJP sum:")
print(float(vjp_sum))

if actual.numpy()[0] == 0.0 and expected[0] != 0.0:
    print("BUG REPRODUCED: tf.nn.log_softmax loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
actual:
[0.0, -5.551115123125776e-17]
expected:
[5.551115123125776e-17, -5.551115123125776e-17]
VJP sum:
-5.551115123125776e-17
BUG REPRODUCED: tf.nn.log_softmax loses a finite first-derivative signal
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
## Summary

- remove the row-sum rounding residual from float64 log-softmax VJPs
- preserve finite tail components when the dominant softmax probability rounds to one
- leave other data types unchanged

Fixes #126630.

## Testing

- `python3 -m py_compile tensorflow/python/ops/nn_grad.py tensorflow/python/kernel_tests/nn_ops/softmax_op_test.py`
- added graph/eager regression coverage for the reported float64 input

### CODE REVIEW

**Comment by GodlyDonuts** [comment]
@gemini-code-assist /review

**Review by dmiltr3 — APPROVED** [review]
Great catch on the float64 residuals and the precision issues. The fix looks mathematically sound and properly preserves tail probabilities safely.

Internal TAP presubmit regression tests have successfully completed. LGTM!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @GodlyDonuts,

Thank you for the contribution!

While syncing this PR with the latest changes on the `master` branch, our automated synchronization tooling encountered merge conflicts in:
- `tensorflow/python/ops/nn_grad.py`
- `tensorflow/python/kernel_tests/nn_ops/BUILD`

These conflicts are due to PR #126748 landing recently on `master`.

Could you please rebase your branch on the latest `master` and resolve the merge conflicts? Once updated, the automated presubmit test suite will run and we can proceed with merging.

Thank you!

**Review by dmiltr3 — APPROVED** [review]
The changes look good and are correct. All internal presubmit tests have passed successfully. The CI failures are due to an unrelated infrastructure issue with LLVM patching on the master branch and can be safely ignored.

Approving!
